### PR TITLE
Add `refill_unblock_check` to many places

### PR DESCRIPTION
**Pull request by michaelmcinerney**
This is some work in progress adding `refill_unblock_check` to many functions, in line with Curtis's `mcs-sporadic-unblock` branch. 

The abstract spec and the Haskell have been updated. I have finished the proofs in AInvs, and have sorried through Refine. I will leave this for the moment to work on removing the `whileM` loops from `refillUnblockCheck` and `refillBudgetCheck` and will come back to it once that's done, so as to not duplicate effort. So please ignore the changes in Refine for the moment.

I have rewritten a fair number of proofs in AInvs which used the backwards reasoning style to instead use forwards reasoning. One example is `send_signal_BOR_helper`. It might be interesting to compare these proofs. Apologies to anyone whose proof I have perhaps unnecessarily rewritten

### CODE REVIEW

File: proof/invariant-abstract/DetSchedSchedule_AI.thy [hunk 1]
```diff
@@ -3647,30 +3650,920 @@ lemma set_thread_state_valid_sched:
                    set_thread_state_valid_sched_action set_thread_state_valid_blocked
                    set_thread_state_released_ipc_queues)
 
+lemma refill_unblock_check_valid_sched_misc[wp]:
```
Comment (michaelmcinerney) [inline proof/invariant-abstract/DetSchedSchedule_AI.thy:3653]: This whole block just moved

File: proof/invariant-abstract/DetSchedSchedule_AI.thy [hunk 2]
```diff
@@ -6529,63 +7418,171 @@ lemma set_tcb_sched_context_released_ipc_queues:
   apply (wpsimp wp: valid_sched_wp)
   by (cases sco; clarsimp simp: is_blocked_thread_state_defs vs_all_heap_simps elim!: released_ipc_queuesE split: if_splits)
 
-lemmas sched_context_resume_st_tcb_at'[wp]
-  = sched_context_resume_st_tcb_at[unfolded obj_at_kh_kheap_simps]
+lemmas sched_context_resume_st_tcb_at'[wp]
```
Comment (michaelmcinerney) [inline proof/invariant-abstract/DetSchedSchedule_AI.thy:7421]: This block mostly just moved

File: proof/invariant-abstract/DetSchedSchedule_AI.thy [hunk 3]
```diff
@@ -6716,69 +7753,137 @@ lemma tcc_valid_sched:
...
+   \<lbrace>\<lambda>_. budget_sufficient t :: det_state \<Rightarrow> _\<rbrace>"
   unfolding install_tcb_cap_def
-  by (wpsimp wp: check_cap_inv cap_delete_fh_lift)
+  apply (wpsimp wp: check_cap_inv hoare_vcg_imp_lift'
+                    cap_delete_fh_lift[where L="active_sc_tcb_at t and released_ipc_queues
+                                                and active_sc_valid_refills
+                                                and current_time_bounded 1"])
+  done
+
+lemma refill_unblock_check_is_refill_ready[wp]:
```
Comment (michaelmcinerney) [inline proof/invariant-abstract/DetSchedSchedule_AI.thy:7798]: The rest of this block just moved

File: proof/invariant-abstract/DetSchedSchedule_AI.thy [hunk 4]
```diff
@@ -12674,128 +13067,309 @@ lemma sc_ready_times_upd:
...
+   postpone sc_ptr
+   \<lbrace>\<lambda>_. not_in_release_q t\<rbrace>"
+  apply (simp add: postpone_def)
+  apply (wpsimp wp: tcb_release_enqueue_not_in_release_q get_sc_obj_ref_wp)
+  by (simp add: sc_at_ppred_def obj_at_def)
+
+crunches postpone
+  for active_sc_at[wp]: "active_sc_at sc_ptr"
+  (wp: crunch_wps)
+
```
Comment (michaelmcinerney) [inline proof/invariant-abstract/DetSchedSchedule_AI.thy:13180]: Before here, this block was just moved

File: proof/invariant-abstract/SchedContext_AI.thy [hunk 5]
```diff
@@ -1029,6 +1029,94 @@ lemma set_tcb_queue_valid_replies[wp]:
   "set_tcb_queue d prio queue \<lbrace> valid_replies_pred P \<rbrace>"
   by (wpsimp simp: set_tcb_queue_def)
 
+lemma update_sched_context_ex_cap_cur_thread [wp]:
```
Comment (michaelmcinerney) [inline proof/invariant-abstract/SchedContext_AI.thy:1032]: This block just moved

File: proof/invariant-abstract/SchedContext_AI.thy [hunk 6]
```diff
@@ -1461,89 +1519,39 @@ lemma possible_switch_to_invs[wp]:
...
-
-lemma refill_unblock_check_state_refs_of_ct[wp]:
-  "\<lbrace>\<lambda>s. P (state_refs_of s) (cur_thread s)\<rbrace>
-     refill_unblock_check scp \<lbrace>\<lambda>_ s. P (state_refs_of s) (cur_thread s)\<rbrace>"
-  apply (wpsimp simp: refill_unblock_check_def is_round_robin_def set_refills_def
-                      update_sched_context_def set_object_def
-                  wp: get_refills_wp get_object_wp)
-  apply (clarsimp simp: state_refs_of_def get_refs_def2 obj_at_def
-                  intro!: ext elim!: rsubst[where P="\<lambda>x. P x (cur_thread s)" for s])
+lemma sched_context_bind_tcb_invs[wp]:
```
Comment (michaelmcinerney) [inline proof/invariant-abstract/SchedContext_AI.thy:1522]: This just moved

**Review by mktnk3 — COMMENTED** [review]
This is a huge chunk of work. Thanks a lot for pushing this to the end of AInvs (and then continuing with Refine). It looks as if you've rewritten most of DetSchedSchedule_AI :) I guess a fair bit of it is due to moving things around, but it must have been quite painful to deal with that file.
Anyway, I am relieved to see that ASpec update/shuffling is not too bad, and that you could finish the AInvs proof update. I hope the haskell/Refine part will come together nicely, too. For now, I've added a few generic comments. I plan to come back to look at the details later bit by bit.

File: proof/refine/ARM/IpcCancel_R.thy [hunk 7]
```diff
@@ -13,8 +13,17 @@ begin
 
 context begin interpretation Arch . (*FIXME: arch_split*)
 
+crunches refillUnblockCheck
```
[the remark this answers is not in the document]
Comment (michaelmcinerney) [reply proof/refine/ARM/IpcCancel_R.thy:16]: I should probably have said to ignore everything in Refine for now, sorry.

Yes we'll definitely need everything for `refillUnblockCheck` in Schedule_R. I'm not sure of any dependency issue there

**Comment by michaelmcinerney** [comment]
Please ignore everything in Refine for now (though it would be good for someone to look at the Haskell)

**Review by corlewis — COMMENTED** [review]
Sorry for the late feedback, but here's a couple of comments after looking through the non DetSchedSchedule_AI files.

In general I agree with Miki. You've clearly put a lot of work into this and you've done a great job!

File: proof/invariant-abstract/IpcCancel_AI.thy [hunk 8]
```diff
@@ -1632,6 +1637,8 @@ end
 
 crunch pred_tcb_at[wp]: empty_slot "\<lambda>s. Q (pred_tcb_at proj P t s)"
 
+crunch sc_at_pred_n[wp]: empty_slot "\<lambda>s. Q (sc_at_pred_n N proj P p s)"
```
Comment (MitchellBuckley) [inline proof/invariant-abstract/IpcCancel_AI.thy:1640]: nitpick: Merge this with previous line using `crunches`?

File: proof/invariant-abstract/IpcDet_AI.thy [hunk 9]
```diff
@@ -1794,20 +1805,26 @@ lemma si_invs'_helper_no_reply:
                                return (sufficient \<and> ready)
               od;
      y <- assert test;
-     y <- set_thread_state dest Running;
-     possible_switch_to dest
+              y <- set_thread_state dest Running;
```
Comment (MitchellBuckley) [inline proof/invariant-abstract/IpcDet_AI.thy:1808]: nitpick: indentation

File: proof/invariant-abstract/IpcDet_AI.thy [hunk 10]
```diff
@@ -1952,11 +1975,11 @@ lemma si_invs'_helper_some_reply:
...
     apply (drule (1) idle_no_ex_cap, clarsimp)
    apply (drule (1) idle_no_ex_cap, clarsimp)
-  apply wpsimp
+  apply (wpsimp wp: hoare_drop_imps)
        apply (strengthen invs_valid_objs, clarsimp cong: conj_cong)
        apply (wpsimp wp: sts_invs_minor2)
-      apply wpsimp
-     apply (wpsimp wp: hoare_drop_imp)
+      apply (wpsimp wp: hoare_drop_imps)
+      apply (wpsimp wp: hoare_drop_imps)
```
Comment (MitchellBuckley) [inline proof/invariant-abstract/IpcDet_AI.thy:1982]: nitpick: check indentation here.

**Review by MitchellBuckley — COMMENTED** [review]
Woah, this is huge. I've made a few comments and AInvs looks pretty good. Let me know when this is really ready and I can look at it again.

**Pull request closed without merging**